File: crates/r4s_engine/src/dice.rs
use std::collections::VecDeque;
use std::hash::{BuildHasher, Hasher};

use crate::error::EngineError;
use crate::types::MAX_DICE_LIMIT;

/// Fuente de d6. Se inyecta para poder tener tiradas deterministas en tests.
pub trait DiceSource {
    /// Devuelve un valor entre 1 y 6.
    fn d6(&mut self) -> u8;
}

/// Generador SplitMix64: rápido, sin dependencias y reproducible con semilla.
#[derive(Debug, Clone)]
pub struct SeededDice {
    state: u64,
}

impl SeededDice {
    pub fn new(seed: u64) -> Self {
        Self { state: seed }
    }

    /// Semilla tomada de la aleatoriedad del proceso (`RandomState` se
    /// inicializa desde el SO) mezclada con la hora actual.
    pub fn from_entropy() -> Self {
        let mut hasher = std::collections::hash_map::RandomState::new().build_hasher();
        let nanos = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or_default();
        hasher.write_u128(nanos);
        Self::new(hasher.finish())
    }

    fn next_u64(&mut self) -> u64 {
        self.state = self.state.wrapping_add(0x9E37_79B9_7F4A_7C15);
        let mut z = self.state;
        z = (z ^ (z >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
        z = (z ^ (z >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
        z ^ (z >> 31)
    }
}

impl DiceSource for SeededDice {
    fn d6(&mut self) -> u8 {
        // Muestreo por rechazo para evitar sesgo de módulo.
        const ZONE: u64 = u64::MAX - (u64::MAX % 6);
        loop {
            let v = self.next_u64();
            if v < ZONE {
                return (v % 6) as u8 + 1;
            }
        }
    }
}

/// Dados guionizados para tests. Entra en pánico si se agotan.
#[derive(Debug, Clone, Default)]
pub struct FixedDice(pub VecDeque<u8>);

impl FixedDice {
    pub fn new(values: impl IntoIterator<Item = u8>) -> Self {
        Self(values.into_iter().collect())
    }
}

impl DiceSource for FixedDice {
    fn d6(&mut self) -> u8 {
        self.0.pop_front().expect("FixedDice sin valores")
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DiceRoll {
    dice: Vec<u8>,
}

impl DiceRoll {
    /// Construye una tirada a partir de valores ya conocidos (p. ej. leídos de Firestore).
    pub fn from_values(dice: Vec<u8>, max_dice: u8) -> Result<Self, EngineError> {
        if dice.is_empty() || dice.len() > max_dice.min(MAX_DICE_LIMIT) as usize {
            return Err(EngineError::DiceCountOutOfRange {
                got: dice.len(),
                max: max_dice,
            });
        }
        if let Some(&bad) = dice.iter().find(|&&d| !(1..=6).contains(&d)) {
            return Err(EngineError::InvalidDieValue(bad));
        }
        Ok(Self { dice })
    }

    pub fn dice(&self) -> &[u8] {
        &self.dice
    }

    pub fn len(&self) -> usize {
        self.dice.len()
    }

    pub fn is_empty(&self) -> bool {
        self.dice.is_empty()
    }

    pub fn total(&self) -> u32 {
        self.dice.iter().map(|&d| u32::from(d)).sum()
    }

    pub fn all_sixes(&self) -> bool {
        self.dice.iter().all(|&d| d == 6)
    }

    pub fn non_sixes(&self) -> u32 {
        self.dice.iter().filter(|&&d| d != 6).count() as u32
    }
}

/// Tira `count` d6.
pub fn roll(count: u8, max_dice: u8, source: &mut impl DiceSource) -> Result<DiceRoll, EngineError> {
    let values = (0..count).map(|_| source.d6()).collect();
    DiceRoll::from_values(values, max_dice)
}

File: web/src/engine/dice.ts
import { EngineError } from './errors';
import { MAX_DICE_LIMIT } from './types';

/** Fuente de d6. Se inyecta para poder tener tiradas deterministas en tests. */
export interface DiceSource {
  /** Devuelve un valor entre 1 y 6. */
  d6(): number;
}

const U64 = (1n << 64n) - 1n;
/** Zona sin sesgo de módulo para 64 bits. */
const ZONE64 = U64 - (U64 % 6n);

/** Generador SplitMix64: rápido, sin dependencias y reproducible con semilla (el mismo que el motor en Rust). */
export class SeededDice implements DiceSource {
  private state: bigint;

  constructor(seed: bigint | number) {
    this.state = BigInt(seed) & U64;
  }

  private nextU64(): bigint {
    this.state = (this.state + 0x9e3779b97f4a7c15n) & U64;
    let z = this.state;
    z = ((z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n) & U64;
    z = ((z ^ (z >> 27n)) * 0x94d049bb133111ebn) & U64;
    return z ^ (z >> 31n);
  }

  d6(): number {
    // Muestreo por rechazo para evitar sesgo de módulo.
    for (;;) {
      const v = this.nextU64();
      if (v < ZONE64) return Number(v % 6n) + 1;
    }
  }
}

/** Zona sin sesgo para 32 bits: 2^32 - (2^32 mod 6). */
const ZONE32 = 0x1_0000_0000 - (0x1_0000_0000 % 6);

/** Dados de verdad: `crypto.getRandomValues` con muestreo por rechazo (sin sesgo de módulo). */
export class CryptoDice implements DiceSource {
  private readonly buf = new Uint32Array(1);

  d6(): number {
    for (;;) {
      crypto.getRandomValues(this.buf);
      const v = this.buf[0]!;
      if (v < ZONE32) return (v % 6) + 1;
    }
  }
}

/** Dados guionizados para tests. Lanza si se agotan. */
export class FixedDice implements DiceSource {
  private readonly values: number[];

  constructor(values: Iterable<number>) {
    this.values = [...values];
  }

  d6(): number {
    const v = this.values.shift();
    if (v === undefined) throw new Error('FixedDice sin valores');
    return v;
  }
}

/** Tirada validada: entre 1 y `maxDice` dados, cada uno entre 1 y 6. Inmutable. */
export class DiceRoll {
  readonly dice: readonly number[];

  private constructor(dice: number[]) {
    this.dice = Object.freeze(dice);
  }

  /** Construye una tirada a partir de valores ya conocidos (p. ej. leídos de Firestore). */
  static fromValues(dice: readonly number[], maxDice: number): DiceRoll {
    if (dice.length === 0 || dice.length > Math.min(maxDice, MAX_DICE_LIMIT)) {
      throw new EngineError({ kind: 'DiceCountOutOfRange', got: dice.length, max: maxDice });
    }
    const bad = dice.find((d) => !Number.isInteger(d) || d < 1 || d > 6);
    if (bad !== undefined) throw new EngineError({ kind: 'InvalidDieValue', value: bad });
    return new DiceRoll([...dice]);
  }

  get length(): number {
    return this.dice.length;
  }

  total(): number {
    return this.dice.reduce((a, d) => a + d, 0);
  }

  allSixes(): boolean {
    return this.dice.every((d) => d === 6);
  }

  nonSixes(): number {
    return this.dice.filter((d) => d !== 6).length;
  }
}

/** Tira `count` d6. */
export function roll(count: number, maxDice: number, source: DiceSource): DiceRoll {
  const values = Array.from({ length: Math.max(0, count) }, () => source.d6());
  return DiceRoll.fromValues(values, maxDice);
}
